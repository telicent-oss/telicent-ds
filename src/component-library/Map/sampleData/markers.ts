export const BOULAY_ATTACK = {
  geohash: "http://geohash.org/sd37f",
  type: "http://acleddata.com/ontology#Attack",
  uri: "http://telicent.io/data#SUD11217",
  name: "Attack in Boulay, Sudan",
};
export const ARGA_ATTACK = {
  geohash: "http://geohash.org/sd30t",
  type: "http://acleddata.com/ontology#Attack",
  uri: "http://telicent.io/data#SUD11675",
  name: "Attack in Arga, Sudan",
};
