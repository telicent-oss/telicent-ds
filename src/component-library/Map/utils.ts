import { maxBy, minBy } from "lodash";
import { LngLatBoundsLike, LngLatLike } from "react-map-gl/maplibre";
import { LocationFeature } from "../../contexts/Location";

type MinMax = "min" | "max";
type LatLng = "latitude" | "longitude";

const getMinOrMax = (
  markers: LocationFeature[],
  minOrMax: MinMax,
  latOrLng: LatLng
) => {
  const getLatLng = (location: LocationFeature) =>
    location.properties[latOrLng];

  let feature = minBy(markers, getLatLng);

  if (minOrMax === "max") {
    feature = maxBy(markers, getLatLng);
  }

  return feature?.properties[latOrLng];
};

export const getBounds = (
  markers: LocationFeature[]
): LngLatBoundsLike | undefined => {
  const maxLat = getMinOrMax(markers, "max", "latitude");
  const minLat = getMinOrMax(markers, "min", "latitude");
  const maxLng = getMinOrMax(markers, "max", "longitude");
  const minLng = getMinOrMax(markers, "min", "longitude");

  if (!minLng || !minLat || !maxLng || !maxLat) return;

  const southWest: LngLatLike = [minLng, minLat];
  const northEast: LngLatLike = [maxLng, maxLat];
  return [southWest, northEast];
};
