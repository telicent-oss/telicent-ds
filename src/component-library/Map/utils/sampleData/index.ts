import { MOLDOVA } from "./moldova.geo";
import { AUSTRIA } from "./austria.geo";

export { MOLDOVA, AUSTRIA };
