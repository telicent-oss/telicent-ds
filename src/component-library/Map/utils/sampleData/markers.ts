export const BOULAY_ATTACK = {
  geohash: "http://geohash.org/sd37f",
  type: "http://acleddata.com/ontology#Attack",
  uri: "http://telicent.io/data#SUD11217",
  name: "Attack in Boulay, Sudan",
};

export const ARGA_ATTACK = {
  geohash: "http://geohash.org/sd30t",
  type: "http://acleddata.com/ontology#Attack",
  uri: "http://telicent.io/data#SUD11675",
  name: "Attack in Arga, Sudan",
};

export const MADE_UP_MARKER = {
  geohash: "http://geohash.org/sd30t",
  type: "http://acleddata.com/ontology#MadeUp",
  uri: "http://telicent.io/data#MUP11675",
  name: "Non event. Not important",
};
