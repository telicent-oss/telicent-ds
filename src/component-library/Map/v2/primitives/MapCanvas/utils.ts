import { Coordinate } from "ol/coordinate";
import Feature from "ol/Feature";
import { Point, Polygon, MultiPolygon } from "ol/geom";
import { fromLonLat } from "ol/proj";
// import Layer from "ol/layer/Layer";
import TileLayer from "ol/layer/Tile";
import { OSM, XYZ } from "ol/source";
import {
  LayerConfig,
  StyleConfig,
  BaseRasterLayerConfig,
  BaseVectorTileLayerConfig,
  OverlayVectorLayerConfig,
  OverlayFeatureConfig,
  LegacyMapConfig,
  LayerMeta,
} from "../../types";
import { StyleLike } from "ol/style/Style";
import GeoJSON from "ol/format/GeoJSON";
import Fill from "ol/style/Fill";
import Stroke from "ol/style/Stroke";
import Style from "ol/style/Style";
import Text from "ol/style/Text";
import { Circle as CircleStyle } from "ol/style";
import { View } from "ol";
import apply from "ol-mapbox-style";
import LayerGroup from "ol/layer/Group";
import VectorTileLayer from "ol/layer/VectorTile";
import VectorTileSource from "ol/source/VectorTile";
import MVT from "ol/format/MVT";
import VectorLayer from "ol/layer/Vector";
import VectorSource from "ol/source/Vector";
import BaseLayer from "ol/layer/Base";

export const defaultVectorStyle = new Style({
  fill: new Fill({ color: "rgba(0, 0, 255, 0.1)" }),
  stroke: new Stroke({ color: "#3399CC", width: 1 }),
  image: new CircleStyle({
    radius: 5,
    fill: new Fill({ color: "#3399CC" }),
    stroke: new Stroke({ color: "#000", width: 1 }),
  }),
  text: undefined,
});

const toOlStyle = (styleConfig: StyleConfig, feature?: unknown): StyleLike => {
  if (!styleConfig) return defaultVectorStyle;

  if (typeof styleConfig === "function") {
    const result = styleConfig(feature);
    return toOlStyle(result);
  }

  return new Style({
    fill: new Fill({ color: styleConfig.fillColor || "rgba(0,0,0,0.1)" }),
    stroke: new Stroke({
      color: styleConfig.strokeColor || "#000",
      width: styleConfig.strokeWidth || 1,
    }),
    image: new CircleStyle({
      radius: styleConfig.radius || 5,
      fill: new Fill({ color: styleConfig.fillColor || "#000" }),
    }),
    text: styleConfig.text ? new Text({ text: styleConfig.text }) : undefined,
  });
};

const getFeatures = (features: OverlayFeatureConfig[]) =>
  features.map((feature) => {
    let geometry;
    switch (feature.type) {
      case "Point":
        geometry = new Point(
          fromLonLat(feature.coordinates as [number, number])
        );
        break;
      // TODO: https://telicent.atlassian.net/browse/TELFE-1327
      // case "Polygon": {
      //   // Ensure the first polygon if coordinates are multi-dimensional
      //   const coords: Coordinate[][] = Array.isArray(feature.coordinates[0][0])
      //     ? (feature.coordinates as Coordinate[][][])[0]
      //     : (feature.coordinates as Coordinate[][]);
      //   geometry = new Polygon(coords);
      //   break;
      // }
      // case "MultiPolygon": {
      //   const coords: Coordinate[][][] = (feature.coordinates as Coordinate[][][]).map(
      //     polygon => polygon.map(ring => ring as Coordinate[])
      //   );
      //   geometry = new MultiPolygon(coords);
      //   break;
      // }
    }
    return new Feature({ geometry, ...feature.meta });
  });

const getBaseVectorTileLayer = (
  layerConfig: BaseVectorTileLayerConfig
): BaseLayer => {
  switch (layerConfig.provider) {
    case "mapbox":
    case "maptiler":
      // use ol-mapbox-style
      const mapLayer = new LayerGroup(); // container layer
      apply(mapLayer, layerConfig.styleUrl || layerConfig.url, {
        accessToken: layerConfig.accessToken,
      });
      return attachMeta(mapLayer, {
        visible: Boolean(layerConfig.visible),
        label: layerConfig.label,
        image: layerConfig.previewImage,
      });

    case "arcgis":
      // OL VectorTile layer for ArcGIS
      const layer = new VectorTileLayer({
        source: new VectorTileSource({
          url: layerConfig.url + "/tile/{z}/{y}/{x}.pbf", // typical ArcGIS endpoint
          format: new MVT(),
        }),
        visible: layerConfig.visible !== false, // what?
      });
      return attachMeta(layer, {
        visible: Boolean(layerConfig.visible),
        label: layerConfig.label,
        image: layerConfig.previewImage,
      });

    case "custom":
      // assume valid style JSON or MVT endpoint; user handles mapping
      const customLayer = new LayerGroup();
      apply(customLayer, layerConfig.styleUrl || layerConfig.url, {
        accessToken: layerConfig.accessToken,
      });
      return attachMeta(customLayer, {
        visible: Boolean(layerConfig.visible),
        label: layerConfig.label,
        image: layerConfig.previewImage,
      });

    default:
      throw new Error(`Unknown vector tile provider: ${layerConfig.provider}`);
  }
};

const getBaseRasterLayer = (layerConfig: BaseRasterLayerConfig) => {
  const layer = new TileLayer({
    visible: layerConfig.visible ?? true,
    source: new XYZ({
      url: layerConfig.url,
    }),
  });

  const { previewImage: image, label, visible } = layerConfig;

  return attachMeta(layer, { image, label, visible: Boolean(visible) });
};

const getOverlayVectorLayer = (
  layerConfig: OverlayVectorLayerConfig
): VectorLayer => {
  return new VectorLayer({
    visible: layerConfig.visible ?? true,
    source: new VectorSource({
      features: [], // TODO: https://telicent.atlassian.net/browse/TELFE-1327
      format: new GeoJSON(),
    }),
    style: layerConfig.style,
  });
};

export const ensureLayers = (layerConfigs: LayerConfig[]): BaseLayer[] => {
  if (!layerConfigs || layerConfigs.length === 0) {
    console.info("No layers provided. Adding default OSM base layer.");
    return [new TileLayer({ source: new OSM() })];
  }

  return layerConfigs.map((layerConfig) => {
    switch (layerConfig.kind) {
      case "base-raster":
        return getBaseRasterLayer(layerConfig);

      case "base-vector-tiles":
        return getBaseVectorTileLayer(layerConfig);

      default:
        throw new Error(`Unknown layer kind: ${JSON.stringify(layerConfig)}`);
    }
  });
};

export const ensureView = (zoom: number, center: Coordinate) => {
  const defaultZoom = 2;
  const defaultCenter: Coordinate = [0, 0];
  if (!zoom) {
    console.info("Zoom not set using default");
  }

  if (!center) {
    console.info("Center coordinates not set, using default");
  }

  return new View({
    zoom: zoom ?? defaultZoom,
    center: fromLonLat(center ?? defaultCenter),
  });
};

export const mapLegacyConfigToLayers = (
  legacy: LegacyMapConfig
): LayerConfig[] => {
  const layers: LayerConfig[] = [];

  if ("vectorStyles" in legacy) {
    layers.push({
      kind: "base-vector-tiles",
      provider: "custom",
      url: legacy.vectorStyles.uri,
      styleUrl: legacy.vectorStyles.uri,
      visible: false,
      previewImage: legacy.vectorStyles.image,
      label: legacy.vectorStyles.label,
    });
  }

  if ("tileSets" in legacy && Array.isArray(legacy.tileSets)) {
    legacy.tileSets.forEach((tileset) => {
      layers.push({
        kind: "base-raster",
        url: tileset.uri,
        attribution: tileset.attribution,
        visible: false,
        previewImage: tileset.image,
        label: tileset.label,
      });
    });
  }

  return layers.map((layer, idx) => ({
    ...layer,
    visible: idx === 0, // set the first layer to be visible
  }));
};

export const attachMeta = (layer: BaseLayer, meta: LayerMeta) => {
  layer.set("meta", meta);
  return layer;
};

export const getMeta = (layer: BaseLayer): LayerMeta => layer.get("meta");
