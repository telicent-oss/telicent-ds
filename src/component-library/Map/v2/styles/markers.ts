import { parseSVG, makeAbsolute } from "svg-path-parser";
import { MarkerStyle, MarkerType } from "../types/markers";
import { Icon, Style } from "ol/style";
import Feature from "ol/Feature";
import Point from "ol/geom/Point";

const iconCache = new Map<string, string>();

// Convert SVG string to data URL
const svgToDataUrl = (svg: string) =>
  `data:image/svg+xml;charset=utf-8,${encodeURIComponent(svg)}`;

// Marker SVG generators
const makePinSvg = ({
  size = 28,
  color = "#FF6600",
  innerSvg,
}: MarkerStyle): string => `
<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}" viewBox="0 0 24 24">
  <path d="M12 2C8 2 5 5 5 9c0 5 7 13 7 13s7-8 7-13c0-4-3-7-7-7z"
        fill="${color}" stroke="${color}" stroke-width="2"/>
    ${innerSvg}
</svg>
`;

/**
 * Extracts a path `d` from a raw SVG string.
 */
const sanitizeFaSvg = (rawSvg: string): string => {
  // If <path> exists, just extract d
  const pathMatch = rawSvg.match(/<path[^>]*d="([^"]+)"/);
  if (pathMatch) return pathMatch[1];

  // If raw d commands exist, use as-is
  const dMatch = rawSvg.match(/M[\s\S]+/);
  if (dMatch) return dMatch[0];

  throw new Error("No path data found in SVG");
};

/**
 * Compute the bounding box of a path `d` string.
 */
const getPathBBox = (d: string) => {
  const commands = makeAbsolute(parseSVG(d));
  let minX = Infinity,
    minY = Infinity,
    maxX = -Infinity,
    maxY = -Infinity;

  commands.forEach((c) => {
    const points: [number, number][] = [];
    if ("x" in c && "y" in c) points.push([c.x, c.y]);
    if ("x1" in c && "y1" in c) points.push([c.x1, c.y1]);
    if ("x2" in c && "y2" in c) points.push([c.x2, c.y2]);
    points.forEach(([x, y]) => {
      if (x != null && y != null) {
        minX = Math.min(minX, x);
        minY = Math.min(minY, y);
        maxX = Math.max(maxX, x);
        maxY = Math.max(maxY, y);
      }
    });
  });

  return { minX, minY, maxX, maxY, width: maxX - minX, height: maxY - minY };
};

/**
 * Scale, center, and color the inner path into a target viewBox.
 */
const scaleAndCenterPath = (
  rawSvg: string,
  color = "#F092D5",
  targetSize = 24,
  scaleFactor = 0.4
) => {
  const d = sanitizeFaSvg(rawSvg);
  const bbox = getPathBBox(d);
  const scale = (targetSize / 512) * scaleFactor;

  // Center of path in original coordinates
  const cx = bbox.minX + bbox.width / 2;
  const cy = bbox.minY + bbox.height / 2;

  const center = targetSize / 2;

  return `<g transform="translate(${center},${center}) scale(${scale}) translate(${-cx},${-cy})">
    <path d="${d}" fill="${color}"/>
  </g>`;
};

export const makeCircleSvg = ({
  size = 28,
  backgroundColor = "#FF6600",
  color = "#fff",
  strokeWidth = 2,
  innerSvg,
}: MarkerStyle): string => {
  const inner = innerSvg ? scaleAndCenterPath(innerSvg, color, 24, 0.4) : "";

  console.log({ inner });
  return `
<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}" viewBox="0 0 24 24">
  <circle cx="12" cy="12" r="10"
          fill="${backgroundColor}" stroke="${color}" stroke-width="${strokeWidth}"/>
  ${inner}
</svg>
`;
};

const makeIconOnlySvg = ({
  size = 28,
  innerSvg,
  color = "#FF6600",
}: MarkerStyle): string => {
  const inner = innerSvg ? scaleAndCenterPath(innerSvg, color, 24, 0.4) : "";

  return `
<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}" viewBox="0 0 24 24" fill="${color}">
    ${inner}
</svg>
`;
};

const iconGenerators: Record<MarkerType, (s: MarkerStyle) => string> = {
  pin: makePinSvg,
  circle: makeCircleSvg,
  icon: makeIconOnlySvg,
};

export const getGeneratedOlIcon = (style: MarkerStyle = {}): Style => {
  const type: MarkerType = style.markerType ?? "pin";
  const inner = style.innerSvg
    ? scaleAndCenterPath(style.innerSvg, style.color, 24, 0.4)
    : "";

  const useFallback = !inner && style.fallbackText;

  let svg: string;

  if (useFallback) {
    // fallback text inside a circle
    svg = `
      <svg xmlns="http://www.w3.org/2000/svg" width="30" height="30" viewBox="0 0 28 28">
        <circle cx="12" cy="12" r="10" fill="${
          style.backgroundColor ?? "#FF6600"
        }" stroke="${style.color ?? "#fff"}" stroke-width="2"/>
        <text x="12" y="12" text-anchor="middle" dominant-baseline="middle" font-size="12" fill="${
          style.color ?? "#fff"
        }">${style.fallbackText}</text>
      </svg>
    `;
  } else {
    // pin/circle/icon SVG with HTML inside
    const generator = iconGenerators[type];
    svg = generator({ ...style, innerSvg: inner });
  }

  const key = `${type}|${inner}|${JSON.stringify(style)}`;
  let url = iconCache.get(key);
  if (!url) {
    url = svgToDataUrl(svg);
    iconCache.set(key, url);
  }

  return new Style({
    image: new Icon({ src: url, scale: 1 }),
  });
};

// Create OL Feature with style
export const createOlMarker = (
  id: string,
  coords: [number, number],
  style: MarkerStyle
): Feature<Point> => {
  const feature = new Feature({
    geometry: new Point(coords),
    id,
  });
  feature.setStyle(getGeneratedOlIcon(style));
  console.log({ feature });
  return feature;
};
