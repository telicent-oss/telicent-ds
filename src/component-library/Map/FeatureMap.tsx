import React, { useEffect, useState } from "react";
import { FeatureCollection, Feature } from "geojson";
import { z } from "zod";
import Map, { Layer, MapProvider, Source, MapRef } from "react-map-gl/maplibre";
import { ErrorBoundary } from "react-error-boundary";
import { StyleOption, ClassIcon } from "./utils/schema";

import ResultsMarkers, { ResultMarker } from "./ResultsMarkers";

import "maplibre-gl/dist/maplibre-gl.css";
import "./map.css";
import { useStyleSelector, MapBoxSourceSchema } from "./layer-selector/useLayerSelector";
import { LayerSelector } from "./layer-selector/LayerSelector";
import { FlexGrid, FlexGridItem, UITheme, UIThemeProvider } from "../../export";
import PolygonMarkers from "./Polygons";
import { calculateBounds } from "./utils/helper";
import { LngLatBounds, type LngLatBoundsLike } from "maplibre-gl";

export const GEOJSON = "geojson";
export const FEATURE_COLLECTION = "FeatureCollection";
const W_H_100 = { height: "100%", width: "100%" };

const initialFeatureCollection: FeatureCollection = {
  type: "FeatureCollection",
  features: [],
};

const initialView = {
  latitude: 53.42148743839479,
  longitude: -3.863068679356047,
  zoom: 3,
  // set max zoom so when you click on a single point the bounds don't zoom in too far.
  maxZoom: 16,
};

export interface FeatureMapProps {
  mapStyleOptions: {
    vectorStyles?: StyleOption | StyleOption[]; // by the looks of it we are only allowed up to one vector style, so why are we accepting an array?
    tileSets?: StyleOption[];
  };
  initialViewState?: typeof initialView;
  markers?: ResultMarker[];
  geoPolygons?: FeatureCollection;
  selected: string[];
  onClickMarker?: (marker: ResultMarker) => void;
  findByClassUri: (maybeClassUri: string) => ClassIcon;
  theme?: UITheme;
  defaultStyle?: string;
  attributionControl?: boolean;
}

const FeatureMap: React.FC<FeatureMapProps> = ({
  mapStyleOptions,
  markers = [],
  selected,
  theme = "DocumentPink",
  geoPolygons = initialFeatureCollection,
  initialViewState = initialView,
  defaultStyle,
  onClickMarker,
  findByClassUri,
  attributionControl = true
}) => {
  const mapContainerRef = React.createRef<MapRef>();
  const [cursor, setCursor] = useState("auto");

  const styleSelector = useStyleSelector(mapStyleOptions);
  const resetCursor = () => {
    setCursor("auto");
  };

  const fitMarkerAndPolygonBounds = () => {
    if (!mapContainerRef.current || (geoPolygons.features.length < 1 && markers.length < 1)) return;
    const map = mapContainerRef.current;

    let bounds: LngLatBounds;
    if (selected.length < 1) {
      bounds = calculateBounds(markers, geoPolygons.features);
    } else {
      const selectedMarkers = markers
        .reduce((acc, marker) => {
          if (selected.find(s => marker.uri === s)) {
            acc.push(marker)
          }
          return acc
        }, [] as ResultMarker[])

      const selectedPolygons = geoPolygons.features.reduce((acc, polygon) => {
        if (selected.find(s => polygon.properties?.iso3166_a3.endsWith(s.split("#")[1]))) {
          acc.push(polygon)
        }
        return acc
      }, [] as Feature[])

      bounds = calculateBounds(selectedMarkers, selectedPolygons)
    }

    map.fitBounds(bounds, { padding: 20, maxZoom: 8, duration: 500 });
  }

  useEffect(() => {
    fitMarkerAndPolygonBounds();
  }, [geoPolygons, markers, selected])

  const onLoad = () => {
    fitMarkerAndPolygonBounds();
  }

  return (
    <UIThemeProvider dark theme={theme}>
      <ErrorBoundary
        fallback={<p>Something went wrong: Failed to load map component</p>}
      >
        <div className="contents">
          <FlexGrid m={0} direction="column" style={W_H_100}>
            <FlexGridItem flexGrow={1}>
              <MapProvider>
                <Map
                  ref={mapContainerRef}
                  onLoad={onLoad}
                  initialViewState={initialViewState}
                  cursor={cursor}
                  id="TelicentMap"
                  interactiveLayerIds={["document-locations-layer"]}
                  scrollZoom
                  mapStyle={defaultStyle}
                  attributionControl={attributionControl}
                  onDragStart={() => setCursor("move")}
                  onDragEnd={resetCursor}
                  onMouseEnter={() => setCursor("pointer")}
                  onMouseLeave={resetCursor}
                >
                  {
                    styleSelector.mapConfig.tileSets.map((el: z.infer<typeof MapBoxSourceSchema>) => (
                      <Source
                        id={`arcgis-raster-${el.label}`}
                        key={`arcgis-raster-${el.label}`}
                        type="raster"
                        tiles={[el.uri]}
                        tileSize={256}
                      >
                        <Layer
                          id={`arcgis-raster-layer-${el.label}`}
                          type="raster"
                          layout={{
                            visibility:
                              styleSelector?.selected?.uri == el.uri
                                ? "visible"
                                : "none",
                          }}
                        />
                      </Source>
                    ))}
                  <ResultsMarkers onClick={onClickMarker} markers={markers} findByClassUri={findByClassUri} selected={selected} />
                  <PolygonMarkers geometryCollection={geoPolygons} />
                </Map>
              </MapProvider>
            </FlexGridItem>
            <FlexGridItem flexGrow={0} p={1}>
              {/* eslint-disable-next-line react/jsx-props-no-spreading */}
              <LayerSelector {...styleSelector.props} />
            </FlexGridItem>
          </FlexGrid>
        </div>
      </ErrorBoundary>
    </UIThemeProvider>
  );
};

export default FeatureMap;
