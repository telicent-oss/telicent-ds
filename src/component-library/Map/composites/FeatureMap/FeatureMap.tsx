// src/components/FeatureMap/FeatureMap.tsx
import React from "react";
import { LayerSelector } from "../../primitives/LayerSelector/composites/LayerSelector/LayerSelector";
import { FlexGrid, FlexGridItem, UIThemeProvider, UITheme } from "../../../../export";
import { ErrorBoundary } from "react-error-boundary";
import { useMapCanvas } from "../../primitives/MapCanvas/useMapCanvas";
import { MapCanvas, MapCanvasProps } from "../../primitives/MapCanvas/MapCanvas";
import { DEFAULT_GEO_POLYGONS, DEFAULT_VIEW } from "../../constants";
import { MapProvider } from "react-map-gl/maplibre";

// TODO bad type
type Picked = Pick<
  MapCanvasProps,
  | "initialViewState"
  | "defaultStyle"
  | "attributionControl"
  | "markers"
  | "geoPolygons"
  | "selected"
  | "onClickMarker"
  | "findByClassUri"
>;

type Optional = Pick<
  Picked,
  "initialViewState" | "geoPolygons" | "attributionControl"
>;

type Optionalized = Partial<Optional>;

type RequiredRest = Omit<Picked, keyof Optional>;

export interface FeatureMapProps
  extends RequiredRest,       // everything except initialViewState & geoPolygons
          Optionalized       // those two, now optional
{
  theme?: UITheme;
  polygonLayers?: (
    | mapboxgl.FillLayer
    | mapboxgl.LineLayer
    | mapboxgl.SymbolLayer
  )[];
}



const FeatureMap: React.FC<FeatureMapProps> = ({
  markers = [],
  selected,
  theme = "DocumentPink",
  geoPolygons = DEFAULT_GEO_POLYGONS,
  initialViewState = DEFAULT_VIEW,
  defaultStyle,
  onClickMarker,
  findByClassUri,
  attributionControl = true,
  polygonLayers,
}) => {
  
  const { mapRef, cursor, onDragStart, onDragEnd, onMouseEnter, onMouseLeave, onLoad } =
    useMapCanvas({ markers, geoPolygons, selected });

  return (
    <UIThemeProvider dark theme={theme}>
      <ErrorBoundary fallback={<p>Failed to load map</p>}>
        <div className="contents">
          <FlexGrid
            m={0}
            direction="column"
            style={{ width: "100%", height: "100%" }}
          >
            <FlexGridItem flexGrow={1}>
              <MapProvider>
                <MapCanvas
                  mapRef={mapRef}
                  initialViewState={initialViewState}
                  cursor={cursor}
                  onDragStart={onDragStart}
                  onDragEnd={onDragEnd}
                  onMouseEnter={onMouseEnter}
                  onMouseLeave={onMouseLeave}
                  onLoad={onLoad}
                  defaultStyle={defaultStyle}
                  attributionControl={attributionControl}
                  markers={markers}
                  geoPolygons={geoPolygons}
                  selected={selected}
                  onClickMarker={onClickMarker}
                  findByClassUri={findByClassUri}
                  polygonLayers={polygonLayers}
                />
              </MapProvider>
            </FlexGridItem>
            <FlexGridItem flexGrow={0} p={1}>
              <LayerSelector />
            </FlexGridItem>
          </FlexGrid>
        </div>
      </ErrorBoundary>
    </UIThemeProvider>
  );
};

export default FeatureMap;
