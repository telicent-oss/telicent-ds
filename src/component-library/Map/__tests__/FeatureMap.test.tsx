import { render, screen } from "@testing-library/react"
import FeatureMap from "../FeatureMap"
import { ClassIcon } from "../utils/schema"
import { ARGA_ATTACK, BOULAY_ATTACK, MADE_UP_MARKER } from "../sampleData/markers"
import { AUSTRIA, MOLDOVA } from "../sampleData"

const findByClassUriMock = (maybeUri: string): ClassIcon => {
  if (maybeUri === ARGA_ATTACK.type) {
    return ({
      faIcon: "bob",
      backgroundColor: "rgb(36, 36, 0)",
      color: "rgb(255,253,4)",
      classUri: maybeUri,
      iconFallbackText: "Arg",
      alt: "test alt text"
    })
  }
  return ({
    backgroundColor: "rgb(18, 18, 18)",
    color: "rgb(221, 221, 221)",
    classUri: MADE_UP_MARKER.type,
    iconFallbackText: "mb",
    alt: "test alt text"
  })
}

describe("FeatureMap", () => {
  it("should render", () => {
    const { container } = render(<FeatureMap theme="DocumentPink" markers={[BOULAY_ATTACK, ARGA_ATTACK]} geoPolygons={{ type: "FeatureCollection", features: [AUSTRIA, MOLDOVA] }} findByClassUri={findByClassUriMock} selected={[]} mapStyleOptions={{}} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
