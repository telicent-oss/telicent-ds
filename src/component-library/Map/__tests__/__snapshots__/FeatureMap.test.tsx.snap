// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureMap should render 1`] = `
<div
  class="contents"
>
  <div
    class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-column css-1s3vbo0-MuiGrid2-root"
    style="height: 100%; width: 100%;"
  >
    <div
      class="MuiGrid2-root MuiGrid2-direction-xs-row css-y6wdth-MuiGrid2-root"
    >
      <div
        id="TelicentMap"
        style="position: relative; width: 100%; height: 100%;"
      />
    </div>
    <div
      class="MuiGrid2-root MuiGrid2-direction-xs-row css-nr30sk-MuiGrid2-root"
    />
  </div>
</div>
`;
